File: src/features/typing/data/go/copy_list_with_random_pointer.go
package main

type Node struct {
	Val    int
	Next   *Node
	Random *Node
}

func copyRandomList(head *Node) *Node {
	if head == nil {
		return nil
	}
	for cur := head; cur != nil; cur = cur.Next.Next {
		cur.Next = &Node{Val: cur.Val, Next: cur.Next}
	}
	for cur := head; cur != nil; cur = cur.Next.Next {
		if cur.Random != nil {
			cur.Next.Random = cur.Random.Next
		}
	}
	dummy := &Node{}
	tail := dummy
	for cur := head; cur != nil; cur = cur.Next.Next {
		tail.Next = cur.Next
		tail = tail.Next
		cur.Next = cur.Next.Next
	}
	return dummy.Next
}

File: src/features/typing/data/dart/copy_list_with_random_pointer.dart
class Node {
  int val;
  Node? next;
  Node? random;
  Node([this.val = 0, this.next, this.random]);
}

class Solution {
  Node? copyRandomList(Node? head) {
    if (head == null) {
      return null;
    }
    for (Node? cur = head; cur != null; cur = cur.next!.next) {
      cur.next = Node(cur.val, cur.next);
    }
    for (Node? cur = head; cur != null; cur = cur.next!.next) {
      if (cur.random != null) {
        cur.next!.random = cur.random!.next;
      }
    }
    Node dummy = Node();
    Node tail = dummy;
    for (Node? cur = head; cur != null; cur = cur.next) {
      Node copy = cur.next!;
      tail.next = copy;
      tail = copy;
      cur.next = copy.next;
    }
    return dummy.next;
  }
}
